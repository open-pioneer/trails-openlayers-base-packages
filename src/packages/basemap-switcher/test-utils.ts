// SPDX-FileCopyrightText: con terra GmbH and contributors
// SPDX-License-Identifier: Apache-2.0
import { MapConfig, MapConfigProvider, MapRegistry } from "@open-pioneer/map";
import TileLayer from "ol/layer/Tile";
import OSM from "ol/source/OSM";
import Stamen from "ol/source/Stamen";
import { createService } from "@open-pioneer/test-utils/services";
import { MapRegistryImpl } from "@open-pioneer/map/MapRegistryImpl";
import { PackageContextProviderProps } from "@open-pioneer/test-utils/react";
import { screen, waitFor } from "@testing-library/react";

// used to avoid a "ResizeObserver is not defined" error
import ResizeObserver from "resize-observer-polyfill";
global.ResizeObserver = ResizeObserver;

export interface SimpleMapOptions {
    center?: { x: number; y: number };
    zoom?: number;
}

export const MAP_ID = "test";

export async function waitForMapMount(parentTestId = "base") {
    return await waitFor(async () => {
        const domElement = await screen.findByTestId(parentTestId);
        const container = domElement.querySelector(".ol-viewport");
        if (!container) {
            throw new Error("map not mounted");
        }
        return domElement;
    });
}
export async function setupMap(options?: SimpleMapOptions) {
    const mapId = "test";
    const mapConfig: MapConfig = {
        initialView: {
            kind: "position",
            center: options?.center ?? { x: 847541, y: 6793584 },
            zoom: options?.zoom ?? 10
        },
        projection: "EPSG:3857",
        layers: [
            {
                id: "b-1",
                title: "OSM",
                isBaseLayer: true,
                visible: false,
                layer: new TileLayer({
                    source: new OSM()
                })
            },
            {
                id: "b-2",
                title: "Toner",
                isBaseLayer: true,
                visible: true,
                layer: new TileLayer({
                    source: new Stamen({ layer: "toner" })
                })
            }
        ]
    };

    const registry = await createService(MapRegistryImpl, {
        references: {
            providers: [new MapConfigProviderImpl(mapId, mapConfig)]
        }
    });

    return { mapId, registry };
}

export function createPackageContextProviderProps(
    service: MapRegistry
): PackageContextProviderProps {
    return {
        services: {
            "map.MapRegistry": service
        }
    };
}

class MapConfigProviderImpl implements MapConfigProvider {
    mapId = "default";
    mapConfig: MapConfig;

    constructor(mapId: string, mapConfig?: MapConfig | undefined) {
        this.mapId = mapId;
        this.mapConfig = mapConfig ?? {};
    }

    getMapConfig(): Promise<MapConfig> {
        return Promise.resolve(this.mapConfig);
    }
}
