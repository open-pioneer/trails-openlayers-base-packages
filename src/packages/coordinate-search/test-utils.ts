// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { act, fireEvent } from "@testing-library/react";

export function showDropdown(projSelect: HTMLElement) {
    // open dropdown to include options in snapshot; react-select creates list of options in dom after opening selection
    act(() => {
        fireEvent.keyDown(projSelect, { key: "ArrowDown" });
    });
}

export function getCurrentOptions() {
    return Array.from(
        document.getElementsByClassName("coordinate-input-select__option")
    ) as HTMLElement[];
}

export function getCurrentOptionValues(options: HTMLElement[]) {
    const values: (string | null)[] = [];
    for (const opt of options) {
        values.push(opt.textContent);
    }
    return values;
}

export function getClearButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const clearButton = buttonDiv.querySelector(".coordinate-input-clear-button");
    if (!clearButton) {
        throw new Error("clear button not rendered");
    }
    return clearButton;
}

export function getCopyButton(coordinateSearchGroup: Element) {
    const buttonDiv = coordinateSearchGroup.querySelector(".chakra-input__right-element");
    if (!buttonDiv) {
        throw new Error("buttons not rendered");
    }
    const copyButton = buttonDiv.querySelector(".coordinate-input-copy-button");
    if (!copyButton) {
        throw new Error("copy button not rendered");
    }
    return copyButton;
}
