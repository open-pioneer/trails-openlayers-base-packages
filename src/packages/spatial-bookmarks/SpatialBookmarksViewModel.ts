// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { MapModel } from "@open-pioneer/map";
import { LocalStorageNamespace, LocalStorageService } from "@open-pioneer/local-storage";
import { v4 as uuid4v } from "uuid";
import { Extent as OlExtent, getCenter } from "ol/extent";
import { transformExtent } from "ol/proj";
import { createLogger } from "@open-pioneer/core";
import { CleanupHandle, reactiveArray, watch } from "@conterra/reactivity-core";

const LOG = createLogger("spatial-bookmark:SpatialBookmarkViewModel");

interface Extent {
    minX: number;
    minY: number;
    maxX: number;
    maxY: number;
}

export interface Bookmark {
    id: string;
    title: string;
    extent: Extent;
    projection: string;
}

export class SpatialBookmarkViewModel {
    private map: MapModel;
    private packageNamespace: LocalStorageNamespace;
    private writableBookmarks = reactiveArray<Bookmark>([]);
    private watchBookmarksHandle: CleanupHandle;

    /**
     * Provides read-only access to the bookmarks array.
     * The UI simply renders the contents of the store and
     * uses the methods below to modify the state.
     */
    get bookmarks(): Bookmark[] {
        return this.writableBookmarks.getItems();
    }

    constructor(map: MapModel, localStorageService: LocalStorageService) {
        this.map = map;
        this.packageNamespace = localStorageService.getNamespace("spatial-bookmarks");

        // Load from local storage on start; save changes whenever bookmarks change.
        this.loadState();
        this.watchBookmarksHandle = watch(
            () => [this.writableBookmarks.getItems()],
            ([bookmarks]) => this.saveState(bookmarks),
            { immediate: false }
        );
    }

    destroy() {
        this.watchBookmarksHandle.destroy();
    }

    /**
     * Creates a new bookmark with the `title` provided by the UI.
     * The current map extent is used for the new bookmark's extent.
     */
    createBookmark(title: string) {
        const olMap = this.map.olMap;
        // minx, miny, maxx, maxy
        const olExtent = olMap.getView().calculateExtent();
        const projection = olMap.getView().getProjection().getCode();
        const extent: Extent = {
            minX: olExtent[0]!,
            minY: olExtent[1]!,
            maxX: olExtent[2]!,
            maxY: olExtent[3]!
        };
        const bookmark: Bookmark = {
            id: uuid4v(),
            title,
            extent,
            projection
        };
        LOG.debug("Created a new bookmark", bookmark);

        const bookmarks = this.writableBookmarks;
        bookmarks.push(bookmark);
    }

    /**
     * Activates the bookmark with the given id by zooming
     * to the bookmark's extent.
     */
    activateBookmark(bookmark: Bookmark) {
        LOG.debug("Activating bookmark", bookmark);
        const extent = this.getBookmarkExtent(bookmark);

        LOG.debug("Attempting to apply extent", extent);
        this.applyExtent(extent);
    }

    /**
     * Deletes the bookmark with the given id.
     */
    deleteBookmark(id: string) {
        LOG.debug("Deleting bookmark", id);
        const index = this.writableBookmarks.findIndex((bookmark) => bookmark.id === id);
        if (index > -1) {
            this.writableBookmarks.splice(index, 1);
        }
    }

    /**
     * Deletes all bookmarks.
     */
    deleteAllBookmarks() {
        LOG.debug("Deleting all bookmarks");
        this.writableBookmarks.splice(0);
    }

    /**
     * Loads the bookmarks from local storage.
     * If the bookmarks are not valid, the initial state (empty array) is restored.
     */
    private loadState() {
        LOG.debug("Restoring bookmarks from local storage");

        const rawBookmarks = this.packageNamespace.get("bookmarks") ?? [];
        try {
            validateBookmarks(rawBookmarks);
            this.writableBookmarks.push(...rawBookmarks);
        } catch (e) {
            LOG.error("Bookmarks data in local storage is invalid, resetting to default value.", e);
            this.writableBookmarks.splice(0);
            this.saveState([]);
        }
    }

    /**
     * Saves the bookmarks to local storage.
     */
    private saveState(bookmarks: Bookmark[]) {
        LOG.debug("Saving bookmarks to local storage");
        this.packageNamespace.set("bookmarks", bookmarks);
    }

    /** Computes an OpenLayers extent for the given bookmark. */
    private getBookmarkExtent(bookmark: Bookmark) {
        const olMap = this.map.olMap;
        const olView = olMap.getView();
        const extent = bookmark.extent;
        const olExtent = [extent.minX, extent.minY, extent.maxX, extent.maxY];
        const viewProjection = olView.getProjection();
        const extentProjection = bookmark.projection;
        return transformExtent(olExtent, extentProjection, viewProjection);
    }

    /** Moves the map to the given extent. */
    private applyExtent(olExtent: OlExtent) {
        const olMap = this.map.olMap;
        const olView = olMap.getView();
        const center = getCenter(olExtent);
        const resolution = olView.getResolutionForExtent(olExtent);
        olView.setCenter(center);
        olView.setResolution(resolution);
    }
}

function validateBookmarks(rawBookmarks: unknown): asserts rawBookmarks is Bookmark[] {
    if (!Array.isArray(rawBookmarks)) {
        throw new Error(`Expected bookmarks from local storage to be an array.`);
    }
    for (const rawBookmark of rawBookmarks) {
        validateBookmark(rawBookmark);
    }
}

function validateBookmark(rawBookmark: unknown): asserts rawBookmark is Bookmark {
    if (!rawBookmark || typeof rawBookmark !== "object") {
        throw new Error(`Expected bookmark from local storage to be an object.`);
    }

    const bookmark = rawBookmark as Record<keyof Bookmark, unknown>;
    if (typeof bookmark.id !== "string") {
        throw new Error("Bookmark does not have a valid id.");
    }
    if (typeof bookmark.title !== "string") {
        throw new Error("Bookmark does not have a valid title.");
    }
    validateExtent(bookmark.extent);
    if (typeof bookmark.projection !== "string") {
        throw new Error("Bookmark does not have a valid projection value.");
    }
}

function validateExtent(rawExtent: unknown): asserts rawExtent is Extent {
    if (!rawExtent || typeof rawExtent !== "object") {
        throw new Error(`Expected bookmark extent from local storage to be an object.`);
    }

    const extent = rawExtent as Record<keyof Extent, unknown>;
    if (
        typeof extent.minX !== "number" ||
        typeof extent.minY !== "number" ||
        typeof extent.maxX !== "number" ||
        typeof extent.maxY !== "number"
    ) {
        throw new Error("Expected bookmark extent from local storage to have valid coordinates.");
    }
}
