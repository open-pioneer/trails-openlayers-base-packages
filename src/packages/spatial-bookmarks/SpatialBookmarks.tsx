// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import {
    Alert,
    AlertIcon,
    Box,
    Button,
    ButtonProps,
    Flex,
    Input,
    List,
    Spacer,
    Text,
    VStack
} from "@open-pioneer/chakra-integration";
import { LocalStorageService } from "@open-pioneer/local-storage";
import { MapModel, useMapModel } from "@open-pioneer/map";
import { CommonComponentProps, useCommonComponentProps } from "@open-pioneer/react-utils";
import { PackageIntl } from "@open-pioneer/runtime";
import classNames from "classnames";
import { Provider as JotaiProvider, useAtomValue } from "jotai";
import { useIntl, useService } from "open-pioneer:react-hooks";
import { FC, KeyboardEvent, ReactNode, useEffect, useRef, useState } from "react";
import { PiMapTrifold, PiTrashSimpleLight } from "react-icons/pi";
import { Bookmark, SpatialBookmarkViewModel } from "./SpatialBookmarksViewModel";

type UIMode = "list" | "create" | "delete";

export interface SpatialBookmarksProps extends CommonComponentProps {
    /**
     * The id of the map.
     */
    mapId: string;
}

/**
 * A component that allows the user to manage a set of spatial bookmarks.
 */
export const SpatialBookmarks: FC<SpatialBookmarksProps> = (props) => {
    const { map } = useMapModel(props.mapId);
    const localStorageService = useService("local-storage.LocalStorageService");
    const viewModel = useViewModel(map, localStorageService);
    return (
        viewModel && (
            // Makes the store accessible to useAtom() etc. in the UI
            <JotaiProvider store={viewModel.store}>
                <SpatialBookmarkUI {...props} viewModel={viewModel} />
            </JotaiProvider>
        )
    );
};

function SpatialBookmarkUI(props: SpatialBookmarksProps & { viewModel: SpatialBookmarkViewModel }) {
    const { viewModel } = props;
    const intl = useIntl();
    const listItemNodes = useRef<HTMLElement[]>([]);
    const [scrollToLastItem, setScrollToLastItem] = useState(false);

    const bookmarks = useAtomValue(viewModel.bookmarks);
    const [bookmarkName, setBookmarkName] = useState<string>("");
    const isValidBookmarkName = bookmarkName.trim().length > 0; // use trim to avoid bookmarks with space character only
    const { containerProps } = useCommonComponentProps("spatial-bookmarks", props);
    const [uiMode, setUiMode] = useState<UIMode>("list");

    const clearBookmarks = () => {
        viewModel.deleteAllBookmarks();
        setUiMode("list");
    };

    const addBookmark = () => {
        if (isValidBookmarkName) {
            viewModel.createBookmark(bookmarkName);
            setUiMode("list");
            setBookmarkName("");
            setScrollToLastItem(true);
        }
    };

    // Scroll to newly created bookmark after render
    useEffect(() => {
        if (scrollToLastItem) {
            const listItems = listItemNodes.current;
            if (listItems) {
                const lastItem = listItems[bookmarks.length - 1];
                lastItem?.scrollIntoView?.();
            }
            setScrollToLastItem(false);
        }
    }, [bookmarks, scrollToLastItem]);

    const deleteContent = () => (
        <VStack>
            <Alert rounded="md" status="warning">
                <AlertIcon />
                {intl.formatMessage({ id: "bookmark.alert.delete" })}
            </Alert>
            <RemoveControls
                intl={intl}
                onClear={clearBookmarks}
                onCancel={() => setUiMode("list")}
            />
        </VStack>
    );

    const createContent = () => (
        <VStack>
            <Alert rounded="md" status="info">
                <AlertIcon />
                {intl.formatMessage({ id: "bookmark.alert.create" })}
            </Alert>
            <Input
                aria-label={intl.formatMessage({ id: "bookmark.input.label" })}
                placeholder={intl.formatMessage({ id: "bookmark.input.placeholder" })}
                value={bookmarkName}
                onChange={(event) => {
                    setBookmarkName(event.target.value);
                }}
                onKeyDown={(e) => {
                    // add bookmark only, if valid bookmark name exists
                    if (e.key === "Enter") {
                        addBookmark();
                    }
                }}
                isRequired
                isInvalid={!isValidBookmarkName}
                autoFocus // eslint-disable-line jsx-a11y/no-autofocus
            />
            <CreateControls
                intl={intl}
                isInvalid={!isValidBookmarkName}
                onCancel={() => {
                    setBookmarkName("");
                    setUiMode("list");
                }}
                onSave={addBookmark}
            />
        </VStack>
    );

    const listContent = () => (
        <>
            {bookmarks.length ? (
                createList(bookmarks, viewModel, intl, listItemNodes)
            ) : (
                <Alert rounded="md" status="info">
                    <AlertIcon />
                    {intl.formatMessage({ id: "bookmark.alert.noSaved" })}
                </Alert>
            )}
            <ListControls
                intl={intl}
                bookmarks={bookmarks}
                showCreate={() => setUiMode("create")}
                showDelete={() => setUiMode("delete")}
            />
        </>
    );

    const content = (
        <>
            {uiMode === "list" && listContent()}
            {uiMode === "create" && createContent()}
            {uiMode === "delete" && deleteContent()}
        </>
    );

    return <Box {...containerProps}>{content}</Box>;
}

function createList(
    bookmarks: Bookmark[],
    viewModel: SpatialBookmarkViewModel,
    intl: PackageIntl,
    listRef: React.MutableRefObject<HTMLElement[]>
) {
    const deleteBtnLabel = intl.formatMessage({
        id: "bookmark.button.deleteOne"
    });

    const bookmarkItems = bookmarks.map((bookmark, idx) => (
        <BookmarkItem
            key={bookmark.id}
            index={idx}
            listItemNodes={listRef}
            bookmark={bookmark}
            onActivate={() => viewModel.activateBookmark(bookmark)}
            onDelete={() => viewModel.deleteBookmark(bookmark.id)}
            deleteBtnLabel={deleteBtnLabel}
        />
    ));

    return (
        <List
            as="ul"
            p={1} // Some padding for children's box shadow
            className="spatial-bookmark-lists"
            listStyleType="none"
            role="listbox"
            spacing={1}
            aria-label={intl.formatMessage({
                id: "bookmark.list.label"
            })}
            overflowY="auto"
            maxHeight="200"
        >
            {bookmarkItems}
        </List>
    );
}

// key name -> index
const ARROW_INDEX_MAP: Record<string, number> = {
    ArrowDown: 1,
    ArrowUp: -1
};

function BookmarkItem(props: {
    index: number;
    listItemNodes: React.MutableRefObject<HTMLElement[]>;
    bookmark: Bookmark;
    onActivate: () => void;
    onDelete: () => void;
    deleteBtnLabel: string;
}): JSX.Element {
    const { index, listItemNodes, bookmark, onDelete, onActivate, deleteBtnLabel } = props;
    const title = bookmark.title;
    const onKeyDown = (evt: KeyboardEvent) => {
        const key = evt.key;
        if (key === "Enter") {
            // Ignore enter from child components (buttons)
            if (evt.target === evt.currentTarget) {
                onActivate();
            }
            return;
        }

        if (ARROW_INDEX_MAP[key] != null) {
            const len = listItemNodes.current?.length;
            if (!len) {
                return;
            }

            let nextIndex = (index + ARROW_INDEX_MAP[key]!) % len;
            if (nextIndex < 0) {
                nextIndex = len - 1;
            }
            listItemNodes.current[nextIndex]?.focus();
        }
    };
    return (
        <Box
            as="li"
            p={1}
            ref={(el: HTMLDivElement | null) => {
                if (!el) {
                    listItemNodes.current.splice(index, 1);
                    return;
                }
                listItemNodes.current[index] = el;
            }}
            className={classNames("spatial-bookmarks-item")}
            tabIndex={0}
            rounded="md"
            role="option"
            cursor="pointer"
            outline={0}
            _hover={{ background: "trails.50" }}
            _focusVisible={{ boxShadow: "outline" }}
            onKeyDown={onKeyDown}
            onClick={onActivate}
        >
            <Flex width="100%" flexDirection="row" align="center" gap={1}>
                <PiMapTrifold />
                <Text ps={2} noOfLines={1}>
                    {title}
                </Text>
                <Spacer />
                <Button
                    className="spatial-bookmarks-item-delete"
                    aria-label={deleteBtnLabel}
                    borderRadius="full"
                    iconSpacing={0}
                    padding={0}
                    colorScheme="red"
                    variant="ghost"
                    leftIcon={<PiTrashSimpleLight />}
                    onClick={(event) => {
                        onDelete();
                        event.stopPropagation();
                    }}
                />
            </Flex>
        </Box>
    );
}

function ListControls(props: {
    intl: PackageIntl;
    bookmarks: Bookmark[];
    showCreate: () => void;
    showDelete: () => void;
}): JSX.Element {
    const { intl, bookmarks, showCreate, showDelete } = props;

    return (
        <ButtonContainer>
            <DialogButton
                isDisabled={bookmarks.length === 0}
                colorScheme="red"
                width={undefined}
                iconSpacing={0}
                leftIcon={<PiTrashSimpleLight />}
                onClick={showDelete}
                aria-label={intl.formatMessage({ id: "bookmark.button.deleteAll" })}
                variant="outline"
            />
            <DialogButton onClick={showCreate} width="100%">
                {intl.formatMessage({ id: "bookmark.button.create" })}
            </DialogButton>
        </ButtonContainer>
    );
}

function RemoveControls(props: { intl: PackageIntl; onClear: () => void; onCancel: () => void }) {
    const { intl, onCancel, onClear } = props;
    return (
        <ButtonContainer>
            <DialogButton variant="outline" onClick={onCancel}>
                {intl.formatMessage({ id: "bookmark.button.cancelDelete" })}
            </DialogButton>
            <DialogButton onClick={onClear}>
                {intl.formatMessage({ id: "bookmark.button.confirmDelete" })}
            </DialogButton>
        </ButtonContainer>
    );
}

function CreateControls(props: {
    intl: PackageIntl;
    isInvalid: boolean;
    onCancel: () => void;
    onSave: () => void;
}) {
    const { intl, onCancel, onSave, isInvalid } = props;
    return (
        <ButtonContainer>
            <DialogButton variant="outline" onClick={() => onCancel()}>
                {intl.formatMessage({ id: "bookmark.button.cancel" })}
            </DialogButton>
            <DialogButton isDisabled={isInvalid} onClick={() => onSave()}>
                {intl.formatMessage({ id: "bookmark.button.save" })}
            </DialogButton>
        </ButtonContainer>
    );
}

function ButtonContainer(props: { children: ReactNode }) {
    return (
        <Flex width="100%" flexDirection="row" mt={2} gap={1}>
            {props.children}
        </Flex>
    );
}

function DialogButton(props?: ButtonProps): JSX.Element {
    return <Button width="100%" {...props} />;
}

function useViewModel(map: MapModel | undefined, localStorageService: LocalStorageService) {
    const [viewModel, setViewModel] = useState<SpatialBookmarkViewModel>();
    useEffect(() => {
        let viewModel: SpatialBookmarkViewModel | undefined;
        if (!map) {
            viewModel = undefined;
        } else {
            viewModel = new SpatialBookmarkViewModel(map, localStorageService);
        }

        setViewModel(viewModel);
        return () => viewModel?.destroy();
    }, [map, localStorageService]);

    return viewModel;
}
