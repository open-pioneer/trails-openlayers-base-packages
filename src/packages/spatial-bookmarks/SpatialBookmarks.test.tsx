// SPDX-FileCopyrightText: 2023 Open Pioneer project (https://github.com/open-pioneer)
// SPDX-License-Identifier: Apache-2.0
import { LocalStorageNamespace, LocalStorageService } from "@open-pioneer/local-storage";
import { createServiceOptions, setupMap } from "@open-pioneer/map-test-utils";
import { PackageContextProvider } from "@open-pioneer/test-utils/react";
import {
    act,
    findByLabelText,
    findByPlaceholderText,
    findByRole,
    findByText,
    render,
    screen
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { afterEach, expect, it } from "vitest";
import { SpatialBookmarks } from "./SpatialBookmarks";
import { Bookmark } from "./SpatialBookmarksViewModel";

const MOCK_NAMESPACE_CONTENT = new Map<string, unknown>();

const MOCK_NAMESPACE = {
    get(key) {
        return MOCK_NAMESPACE_CONTENT.get(key);
    },
    set(key, value) {
        MOCK_NAMESPACE_CONTENT.set(key, value);
    }
} satisfies Partial<LocalStorageNamespace>;

const MOCK_STORAGE_SERVICE: Partial<LocalStorageService> = {
    getNamespace(key): LocalStorageNamespace {
        if (key !== "spatial-bookmarks") {
            throw new Error("unexpected key");
        }

        return MOCK_NAMESPACE as LocalStorageNamespace;
    }
};

afterEach(() => {
    MOCK_NAMESPACE_CONTENT.clear();
});

it("should shown an alert message if no bookmarks are saved", async () => {
    MOCK_NAMESPACE.set("bookmarks", []);

    const { div } = await createBookmarkComponent();
    const alertDiv = await findByRole(div, "alert");
    expect(alertDiv).toMatchSnapshot();
});

it("should load the stored bookmark from local storage and present it visibly within the list", async () => {
    createBookmarks();
    const { div } = await createBookmarkComponent();
    const bookmarkList = await findByRole(div, "listbox");
    expect(bookmarkList).toMatchSnapshot();
});

it("should save a bookmark and update the list", async () => {
    const user = userEvent.setup();
    MOCK_NAMESPACE.set("bookmarks", []);

    const { div } = await createBookmarkComponent();

    const createBtn = await findByText(div, "bookmark.button.create");

    await user.click(createBtn);
    const input = await findByPlaceholderText(div, "bookmark.input.placeholder");
    const saveBtn = await findByText(div, "bookmark.button.save");

    await user.type(input, "Dortmund");
    await user.click(saveBtn);
    const bookmarks = MOCK_NAMESPACE.get("bookmarks") as Bookmark[];

    expect(bookmarks[0]?.title).toEqual("Dortmund");
});

it("should only delete all bookmarks after the user confirms the action.", async () => {
    const user = userEvent.setup();

    createBookmarks();
    const { div } = await createBookmarkComponent();

    const deleteAllBtn = await findByLabelText(div, "bookmark.button.deleteAll");
    await user.click(deleteAllBtn);
    const alertMessage = await findByRole(div, "alert");
    expect(alertMessage).toMatchSnapshot();

    const confirmBtn = await findByText(div, "bookmark.button.confirmDelete");
    await user.click(confirmBtn);
    const noSavedBookmarksAlert = await findByText(div, "bookmark.alert.noSaved");
    expect(noSavedBookmarksAlert).toMatchSnapshot();

    const bookmarks = MOCK_NAMESPACE.get("bookmarks") as Bookmark[];
    expect(bookmarks).toHaveLength(0);
});

it("should remove bookmark from the list by clicking the remove button", async () => {
    const user = userEvent.setup();

    createBookmarks();
    const { div } = await createBookmarkComponent();

    const deleteBookmarkBtn = await findByLabelText(div, "bookmark.button.deleteOne");
    await user.click(deleteBookmarkBtn);
    const noSavedBookmarksAlert = await findByText(div, "bookmark.alert.noSaved");
    expect(noSavedBookmarksAlert).toMatchSnapshot();

    const bookmarks = MOCK_NAMESPACE.get("bookmarks") as Bookmark[];
    expect(bookmarks).toHaveLength(0);
});

it("should center the map on the selected bookmark", async () => {
    const user = userEvent.setup();
    const bookmarks = createBookmarks();
    const { mapModel, div } = await createBookmarkComponent();

    const extent = bookmarks[0]!.extent;
    const bookmarkExtent = [extent.minX, extent.minY, extent.maxX, extent.maxY];

    const initExtent = mapModel.olMap.getView().calculateExtent();
    expect(initExtent).not.toStrictEqual(bookmarkExtent);

    const bookmarkItem = await findByText(div, "Test");
    await user.click(bookmarkItem);
    const newExtent = mapModel.olMap.getView().calculateExtent();
    expect(newExtent).toStrictEqual(bookmarkExtent);
});

it("should move the focus up and down inside the bookmarks list", async () => {
    const user = userEvent.setup();
    createBookmarks(2);
    const { div } = await createBookmarkComponent();

    const listItems = div.querySelectorAll("li");
    expect(listItems).toHaveLength(2);

    const firstItem = listItems[0]!;
    const secondItem = listItems[1]!;

    // Init focus to first item
    act(() => firstItem.focus());
    expect(document.activeElement).toBe(firstItem);

    // Downarrow focuses second item
    await user.keyboard("{ArrowDown}");
    expect(document.activeElement).toBe(secondItem);

    // Wraps to first item
    await user.keyboard("{ArrowDown}");
    expect(document.activeElement).toBe(firstItem);

    // Up wraps back to second item
    await user.keyboard("{ArrowUp}");
    expect(document.activeElement).toBe(secondItem);

    // ... and up to first
    await user.keyboard("{ArrowUp}");
    expect(document.activeElement).toBe(firstItem);
});

interface CtxProviderConfig {
    "local-storage.LocalStorageService": Partial<LocalStorageService>;
}

interface ComponentPropsConfig {
    mapId: string;
    "data-testid": string;
}

async function createBookmarkComponent() {
    const { mapId, mapModel, injectedServices } = await setupComponent();
    renderComponent({ services: injectedServices }, { mapId: mapId, "data-testid": "bookmarks" });
    const div = await waitForSpatialBookmarks();
    return { mapId, mapModel, div };
}

async function setupComponent() {
    const { mapId, registry } = await setupMap({
        projection: "EPSG:25832"
    });
    const mapModel = await registry.expectMapModel(mapId);
    const injectedServices = {
        ...createServiceOptions({ registry }),
        "local-storage.LocalStorageService": MOCK_STORAGE_SERVICE
    };
    return {
        mapModel,
        mapId,
        injectedServices
    };
}
function renderComponent(
    contextProviderProps: { services: CtxProviderConfig },
    componentProps: ComponentPropsConfig
) {
    return render(
        <PackageContextProvider {...contextProviderProps}>
            <SpatialBookmarks {...componentProps} />
        </PackageContextProvider>
    );
}
async function waitForSpatialBookmarks() {
    return await screen.findByTestId<HTMLDivElement>("bookmarks");
}

function createBookmarks(count = 1) {
    const bookmarks: Bookmark[] = [];
    for (let i = 0; i < count; ++i) {
        const bookmark: Bookmark = {
            id: `test-${i}`,
            title: "Test",
            extent: {
                minX: 404609.0492711461,
                minY: 5759938.978042119,
                maxX: 405066.2854485374,
                maxY: 5760396.21421951
            },
            projection: "EPSG:25832"
        };
        bookmarks.push(bookmark);
    }

    MOCK_NAMESPACE.set("bookmarks", bookmarks);
    return bookmarks;
}
