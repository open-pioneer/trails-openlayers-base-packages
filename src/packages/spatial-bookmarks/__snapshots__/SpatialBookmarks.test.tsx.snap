// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should load the stored bookmark from local storage and present it visibly within the list 1`] = `
<ul
  aria-label="bookmark.list.label"
  class="spatial-bookmark-list chakra-list__root css-1q4ghp7"
  role="listbox"
>
  <li
    class="spatial-bookmarks-item chakra-list__item css-843j6g"
    role="option"
    tabindex="0"
  >
    <div
      class="css-v3qtod"
    >
      <div
        class="css-d82rsv"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M228.92,49.69a8,8,0,0,0-6.86-1.45L160.93,63.52,99.58,32.84a8,8,0,0,0-5.52-.6l-64,16A8,8,0,0,0,24,56V200a8,8,0,0,0,9.94,7.76l61.13-15.28,61.35,30.68A8.15,8.15,0,0,0,160,224a8,8,0,0,0,1.94-.24l64-16A8,8,0,0,0,232,200V56A8,8,0,0,0,228.92,49.69ZM104,52.94l48,24V203.06l-48-24ZM40,62.25l48-12v127.5l-48,12Zm176,131.5-48,12V78.25l48-12Z"
          />
        </svg>
      </div>
      <p
        class="css-29in21"
      >
        Test
      </p>
      <div
        class="css-xsawf1"
      />
      <button
        aria-label="bookmark.button.deleteOne"
        class="chakra-button chakra-tooltip__trigger spatial-bookmarks-item-delete css-1qwewjl"
        data-part="trigger"
        data-scope="tooltip"
        data-state="closed"
        dir="ltr"
        id="tooltip:«r0»:trigger"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 256 256"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M216,50H40a6,6,0,0,0,0,12H50V208a14,14,0,0,0,14,14H192a14,14,0,0,0,14-14V62h10a6,6,0,0,0,0-12ZM194,208a2,2,0,0,1-2,2H64a2,2,0,0,1-2-2V62H194ZM82,24a6,6,0,0,1,6-6h80a6,6,0,0,1,0,12H88A6,6,0,0,1,82,24Z"
          />
        </svg>
      </button>
    </div>
  </li>
</ul>
`;

exports[`should only delete all bookmarks after the user confirms the action. 1`] = `
<div
  class="chakra-alert__root css-jwwdmt"
  role="alert"
>
  <span
    class="css-13gnd3a"
  >
    <svg
      class="css-126y6n1"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11 15H13V17H11V15ZM11 7H13V13H11V7Z"
      />
    </svg>
  </span>
  <div
    class="chakra-alert__title css-1jeznpc"
  >
    bookmark.alert.delete
  </div>
</div>
`;

exports[`should only delete all bookmarks after the user confirms the action. 2`] = `
<div
  class="chakra-alert__title css-1jeznpc"
>
  bookmark.alert.noSaved
</div>
`;

exports[`should remove bookmark from the list by clicking the remove button 1`] = `
<div
  class="chakra-alert__title css-1jeznpc"
>
  bookmark.alert.noSaved
</div>
`;

exports[`should shown an alert message if no bookmarks are saved 1`] = `
<div
  class="chakra-alert__root css-1owmyes"
  role="alert"
>
  <span
    class="css-13gnd3a"
  >
    <svg
      class="css-1g7m1fs"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22ZM12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20ZM11 7H13V9H11V7ZM11 11H13V17H11V11Z"
      />
    </svg>
  </span>
  <div
    class="chakra-alert__title css-1jeznpc"
  >
    bookmark.alert.noSaved
  </div>
</div>
`;
